### ISSUE TITLE
Visual explainability of object detection via D-rise implementation

### ISSUE THREAD

**Issue by victor-flindt** [ultralytics/yolov5#9935]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar feature requests.

### Description

As a step towards more explainable AI, i think it would make sense to implement some form of visual explainability of the decision making of the network, rather this be through methods such as Gradcam or RISE /D-RISE. 

### Use case

This would help make the model less of a blackbox to Management types, and would certianly be a great tool for Data Engineers and Scientist to debug unexpected behavior / predictions. 

Also this would help for scenarios where a model might have great acc/recall but further explanation is needed due to government or internal standards.  

### Additional

I currently have a working Notebook implementation based on the article https://arxiv.org/abs/2006.03204 and https://github.com/hysts/pytorch_D-RISE. 

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@victor-flindt hi thanks for the issue! The D-Rise visuals look very interesting, but since it's only in mmdetection there's no easy implementation avenue here. Do you have a YOLOv5 implementation you might be able to share or even better submit a PR for?

The closest we have right now is simply feature visualization. Here's a few details.

### YOLOv5 Feature Visualization

YOLOv5 🚀 features can be visualized through all stages of the model from input to output. To visualize features from a given source run `detect.py` with the `--visualize` flag:
```bash
python detect.py --weights yolov5s.pt --source data/images/bus.jpg --visualize
```

An example [Notebook](https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb) visualizing [bus.jpg](https://github.com/ultralytics/yolov5/blob/master/data/images/bus.jpg) features with YOLOv5s is shown below:

<a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
<img width="1811" alt="Screenshot 2021-08-30 at 16 44 04" src="https://user-images.githubusercontent.com/26833433/131357928-663a120a-3133-4e1a-8995-c812bd732e8a.png">

All stages are visualized by default, each with its own PNG showing the first 32 feature maps **output** from that stage. You can open any PNG for a closer look. For example the first 32 feature maps of the Focus() layer output are shown in `stage0_Focus_features.png`:

![stage0_Focus_features](https://user-images.githubusercontent.com/26833433/131359874-ecc9aef8-1c9b-44f5-ac17-d6b987093554.png)

Feature maps may be customized by updating the `feature_visualization()` function in `utils/plots.py`:
https://github.com/ultralytics/yolov5/blob/bb5ebc290e5d630a081d7cbc5a9725ed8cea0a24/utils/plots.py#L403-L427

Good luck 🍀 and let us know if you have any other questions!

**Comment by Zephyr69** [comment]
There is a grad-cam implementation post for v5.0. You may check it out.

**Comment by glenn-jocher** [comment]
@Zephyr69 thank you for the heads up! The Grad-CAM implementation is indeed available in the [YOLOv5 V5.0 release](https://docs.ultralytics.com/yolov5/releases), and while it may require adjustments for the latest V6.0, it still provides a good starting point for visual explainability. Your feedback is appreciated, and we're constantly working to improve and expand the capabilities of YOLOv5. Let us know if you have any further questions or ideas!
